Initial Symbolic Shape Analysis

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56966 [JIT] Disable Complete Shape Inlining For Testing Purposes
* #56828 Add unary/binary ops necessary for mobilenet
* #56642 Only optimize while change has been made
* #56438 [JIT] integer value refinement
* #56437 Factor out isDominatedBy
* #55978 Add Value * == Value * peephole
* #55977 [JIT] [Easy] Factor out non tensor peephole
* #55926 Add list len refinement
* #55925 Add handling of symbolic shapes
* **#54809 Initial Symbolic Shape Analysis**

I'm going to post on dev-discuss soon with a more thorough explanation of the design and advantages of this shape analysis, so I'm leaving out that for now.

There is still a ton left to do, I'm posting this initial version so we can get something on master multiple can work on. List of many remaining steps to do:

- [ ] Add symbolic shapes support
- [ ] Bind shape functions for operators in C++ 
- [ ] Make classes of operators share the same shape function (e.g. pointwise, broadcast two inputs)
- [ ] Refactor APIs 
- [ ] Only iteratively optimize shape function while a change has been made
- [ ] Expand coverage of coverage to common ops
- [ ] Add shape analysis pass on Graph that handles Ifs and Loops 
- [ ] Allow concurrent reads to the operator map
- [ ] Successive applications of same inputs to same shape function (e.g. series of pointwise ops)

For this review, I am mostly looking for comments related to the implementation of symolic_shape_analysis.cpp, with the caveats listed above. I am not really looking for comments related to api/registration/graph level analysis as those are all planned to be changed. I am fine landing this as is or waiting until necessary components of the TODOs above are finished.

Differential Revision: [D27750998](https://our.internmc.facebook.com/intern/diff/D27750998)

Commit:
Initial Symbolic Shape Analysis

[ghstack-poisoned]

Commit:
Update on "Initial Symbolic Shape Analysis"


I'm going to post on dev-discuss soon with a more thorough explanation of the design and advantages of this shape analysis, so I'm leaving out that for now.

There is still a ton left to do, I'm posting this initial version so we can get something on master multiple can work on. List of many remaining steps to do:

- [ ] Add symbolic shapes support
- [ ] Bind shape functions for operators in C++ 
- [ ] Make classes of operators share the same shape function (e.g. pointwise, broadcast two inputs)
- [ ] Refactor APIs 
- [ ] Only iteratively optimize shape function while a change has been made
- [ ] Expand coverage of coverage to common ops
- [ ] Add shape analysis pass on Graph that handles Ifs and Loops 
- [ ] Allow concurrent reads to the operator map
- [ ] Successive applications of same inputs to same shape function (e.g. series of pointwise ops)

For this review, I am mostly looking for comments related to the implementation of symolic_shape_analysis.cpp, with the caveats listed above. I am not really looking for comments related to api/registration/graph level analysis as those are all planned to be changed. I am fine landing this as is or waiting until necessary components of the TODOs above are finished. 

[ghstack-poisoned]

penguinwu commented on test/jit/test_symbolic_shape_analysis.py:
```diff
@@ -0,0 +1,55 @@
+class TestSymbolicShapeAnalysis(JitTestCase):
+    def test_shape_analysis(self):
+        @torch.jit.script
+        def broadcast(a: List[int], b: List[int]):
```
Since the test cases use `[None, None]` to set input sizes. Should the parameters of shape functions be `List[Optional[int]]`?

eellison replied:
Optional here is sugar around creating a new symbolic shape (i.e., not knowing anything about it). It's taken from the withSymbolicShapes API. By the time the function is invoked  the Nones get bound to an integer

penguinwu commented on torch/csrc/jit/passes/loop_unrolling.h:
```diff
@@ -7,6 +7,9 @@ namespace jit {
 
 TORCH_API void UnrollLoops(std::shared_ptr<Graph>& graph);
 
+// Only unrolls constant loops. Will unroll them regardless of loop block size
```
When `constant_only` is true, UnrollLoop seems to unroll both small-block loops and constant loops (regardless of sizes). The comment states "Only unrolls constant loops", which is confusing.

penguinwu commented on torch/csrc/jit/passes/loop_unrolling.cpp:
```diff
@@ -162,10 +162,11 @@ void replaceLoopCounter(Node* loop) {
 }
 
-void unroll(Node* loop) {
+void unroll(Node* loop, bool constant_only) {
```
Add a comment on what `constant_only` means.

[Readability] "constant_only" => "ignore_size_limit_for_constant_loop" (or something to that effect, I just feel that "constant_only" is misleading as we are unrolling both small-size loops and constant loops)

penguinwu commented on torch/csrc/jit/passes/loop_unrolling.cpp:
```diff
@@ -162,10 +162,11 @@ void replaceLoopCounter(Node* loop) {
   Block* body = loop->blocks().at(0);
-  if (!isSmallBlock(body))
+
+  if (!isSmallBlock(body) && !constant_only)
```
[Readability issue] This logic is a bit confusing. If I only read the implementation of `unroll()`, it seems to unroll both small-block loops and constant loops (regardless of sizes). But then the name `constant_only` is confusing.

But If I read the invocation contact of unroll, it is clear that when `constant_only` is true, we only unroll constant loops.
```
     if (constant_only &&
           node->inputs().at(0)->node()->kind() != prim::Constant) {
         continue;
       }
      unroll(node, constant_only);
```

For better readability, we either push the checking for `prim::Constant` into `unroll()`; or move the checking logic out of `unroll()` and remove `constant_only` parameter to `unroll()` so that `unroll()` just unroll a loop (all the unrolling decision-making is done before calling `unroll()`.

penguinwu reviewed (approved):
I learned quite a bit from this code (though cannot say I understood 100% of the details). Thanks for putting the initial PR in. Looks good to me, I just have one minor readability issue.

Commit:
Update on "Initial Symbolic Shape Analysis"


I'm going to post on dev-discuss soon with a more thorough explanation of the design and advantages of this shape analysis, so I'm leaving out that for now.

There is still a ton left to do, I'm posting this initial version so we can get something on master multiple can work on. List of many remaining steps to do:

- [ ] Add symbolic shapes support
- [ ] Bind shape functions for operators in C++ 
- [ ] Make classes of operators share the same shape function (e.g. pointwise, broadcast two inputs)
- [ ] Refactor APIs 
- [ ] Only iteratively optimize shape function while a change has been made
- [ ] Expand coverage of coverage to common ops
- [ ] Add shape analysis pass on Graph that handles Ifs and Loops 
- [ ] Allow concurrent reads to the operator map
- [ ] Successive applications of same inputs to same shape function (e.g. series of pointwise ops)

For this review, I am mostly looking for comments related to the implementation of symolic_shape_analysis.cpp, with the caveats listed above. I am not really looking for comments related to api/registration/graph level analysis as those are all planned to be changed. I am fine landing this as is or waiting until necessary components of the TODOs above are finished. 

[ghstack-poisoned]

Commit:
Update on "Initial Symbolic Shape Analysis"


I'm going to post on dev-discuss soon with a more thorough explanation of the design and advantages of this shape analysis, so I'm leaving out that for now.

There is still a ton left to do, I'm posting this initial version so we can get something on master multiple can work on. List of many remaining steps to do:

- [ ] Add symbolic shapes support
- [ ] Bind shape functions for operators in C++ 
- [ ] Make classes of operators share the same shape function (e.g. pointwise, broadcast two inputs)
- [ ] Refactor APIs 
- [ ] Only iteratively optimize shape function while a change has been made
- [ ] Expand coverage of coverage to common ops
- [ ] Add shape analysis pass on Graph that handles Ifs and Loops 
- [ ] Allow concurrent reads to the operator map
- [ ] Successive applications of same inputs to same shape function (e.g. series of pointwise ops)

For this review, I am mostly looking for comments related to the implementation of symolic_shape_analysis.cpp, with the caveats listed above. I am not really looking for comments related to api/registration/graph level analysis as those are all planned to be changed. I am fine landing this as is or waiting until necessary components of the TODOs above are finished.

Differential Revision: [D27750998](https://our.internmc.facebook.com/intern/diff/D27750998)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -0,0 +1,202 @@
+};
+
+void PropagateShapesWithShapeFunction(
+    Node* n,
+    const std::shared_ptr<Graph>& graph) {
+  c10::SymbolicShape out = SymbolicShapeAnalyzer(graph, n).run();
```
Nit: a minor readability complaint: arguments in the function are node-graph, but arguments in the constructor are graph-node (swapped). Also, when I see "graph", by default I think of the "main" graph which "n" belongs to, while in fact this is a shape-function graph - it might be worth reflecting that in the variable name.

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -0,0 +1,202 @@
+
+  c10::SymbolicShape run() {
+    // TODO: only run while the last iteration has made a change
+    for (size_t i = 0; i < 6; i++) {
```
I kind of agree with clang-tidy on that one :D

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -0,0 +1,202 @@
+    Node* list_construct = output->node();
+    std::vector<c10::optional<int64_t>> output_shape;
+    for (Value* input : list_construct->inputs()) {
+      output_shape.push_back(constant_as<int64_t>(input));
```
Does `constant_as` throw in case the value is not constant? Should we add an explicit check for that?

eellison replied:
It returns c10::optional<int64_t>

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -0,0 +1,202 @@
+  }
+
+  void substituteTensorProperties(int64_t tensor_index) {
+    auto tensor_value = node_->inputs().at(tensor_index);
```
Nit: could we use `->input(tensor_index)` here?

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -0,0 +1,202 @@
+    }
+  }
+
+  void substituteTensorProperties(int64_t tensor_index) {
```
Nit: I find it a bit confusing that we use different terms here and there without defining what they mean: tensor_indices, dims, shape, tensor_value. What's the difference between all of them? Could we add some comments?

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.h:
```diff
@@ -0,0 +1,22 @@
+
+// CAUTION NOT TO BE USED, STILL A WIP, NOT STABLE
+
+TORCH_API void PropagateShapesWithShapeFunction(
+    Node* n,
+    const std::shared_ptr<Graph>& graph);
```
I know you asked not to comment on the API, but for my better understanding - do we want to keep that as an external API?

eellison replied:
This is just an API for testing really I can delete this API, i don't think it's used at the moment. Eventually you will register shape functions for particular schemas and then we will have an API for propagating shapes on a graph or node, so users should never be directly passing in a shape function.

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.h:
```diff
@@ -0,0 +1,22 @@
+    Node* n,
+    const std::shared_ptr<Graph>& graph);
+
+TORCH_API void PropagateShapesOnGraph(std::shared_ptr<Graph>& graph);
```
If I were to use this API now (I actually am going to for the aot pipeline), am I expected to 1) set types for the graph inputs, 2) call this function?

eellison replied:
Yea, exactly, although it should still work reasonably well even without input types (without complete shape information obviously).

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.h:
```diff
@@ -0,0 +1,22 @@
+
+TORCH_API void RegisterOperatorShapeFunction(
+    Node* n,
+    const std::shared_ptr<Graph>& graph);
```
Nit: we might be more consistent about how we pass shared pointers. In this PR in different places they are passed by value, by reference, and by constant reference.

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -0,0 +1,202 @@
+#include <torch/csrc/jit/runtime/operator.h>
+#include <torch/csrc/jit/serialization/python_print.h>
+#include <torch/csrc/utils/memory.h>
+#include <exception>
+#include <iostream>
```
These headers don't seem to be used.

ZolotukhinM reviewed (approved):
Overall looks good! I left some minor remarks.

Commit:
Update on "Initial Symbolic Shape Analysis"


I'm going to post on dev-discuss soon with a more thorough explanation of the design and advantages of this shape analysis, so I'm leaving out that for now.

There is still a ton left to do, I'm posting this initial version so we can get something on master multiple can work on. List of many remaining steps to do:

- [ ] Add symbolic shapes support
- [ ] Bind shape functions for operators in C++ 
- [ ] Make classes of operators share the same shape function (e.g. pointwise, broadcast two inputs)
- [ ] Refactor APIs 
- [ ] Only iteratively optimize shape function while a change has been made
- [ ] Expand coverage of coverage to common ops
- [ ] Add shape analysis pass on Graph that handles Ifs and Loops 
- [ ] Allow concurrent reads to the operator map
- [ ] Successive applications of same inputs to same shape function (e.g. series of pointwise ops)

For this review, I am mostly looking for comments related to the implementation of symolic_shape_analysis.cpp, with the caveats listed above. I am not really looking for comments related to api/registration/graph level analysis as those are all planned to be changed. I am fine landing this as is or waiting until necessary components of the TODOs above are finished.

Differential Revision: [D27750998](https://our.internmc.facebook.com/intern/diff/D27750998)

[ghstack-poisoned]

Commit:
Update on "Initial Symbolic Shape Analysis"


I'm going to post on dev-discuss soon with a more thorough explanation of the design and advantages of this shape analysis, so I'm leaving out that for now.

There is still a ton left to do, I'm posting this initial version so we can get something on master multiple can work on. List of many remaining steps to do:

- [ ] Add symbolic shapes support
- [ ] Bind shape functions for operators in C++ 
- [ ] Make classes of operators share the same shape function (e.g. pointwise, broadcast two inputs)
- [ ] Refactor APIs 
- [ ] Only iteratively optimize shape function while a change has been made
- [ ] Expand coverage of coverage to common ops
- [ ] Add shape analysis pass on Graph that handles Ifs and Loops 
- [ ] Allow concurrent reads to the operator map
- [ ] Successive applications of same inputs to same shape function (e.g. series of pointwise ops)

For this review, I am mostly looking for comments related to the implementation of symolic_shape_analysis.cpp, with the caveats listed above. I am not really looking for comments related to api/registration/graph level analysis as those are all planned to be changed. I am fine landing this as is or waiting until necessary components of the TODOs above are finished.

Differential Revision: [D27750998](https://our.internmc.facebook.com/intern/diff/D27750998)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -0,0 +1,208 @@
+// infer that the 4th dimension will have the same symbolic shape as inp[3]
+
+struct SymbolicShapeAnalyzer {
+  SymbolicShapeAnalyzer(Node* n, std::shared_ptr<Graph> shape_compute_graph)
```
Nit: `std::shared_ptr<Graph>&`

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -0,0 +1,208 @@
+  // and non-fixed dimensions we cannot reason about at all
+  // TODO: might be cleaner to store as a pair of index -> symbolic shape
+  // but there were weird lifetime issues
+  std::vector<int64_t> node_input_tensor_indices;
```
Nit: a private member should have underscore at the end.

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -0,0 +1,208 @@
+
+ private:
+  void substituteInputTensorProperties() {
+    for (auto index : node_input_tensor_indices) {
```
Do we really need `node_input_tensor_indices`? Can we go through all the indexes and just add a check from lines 61-70 to skip inputs that are not Tensors with symbolic shapes?

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -0,0 +1,208 @@
+              graph_->inputs().at(i), *tt->sizes().concrete_sizes());
+          continue;
+        }
+        // we can
```
![image](https://user-images.githubusercontent.com/4588111/115487661-00b47380-a20e-11eb-830c-f0e7c2454e85.png)

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -0,0 +1,208 @@
+
+  c10::SymbolicShape run() {
+    // TODO: only run while the last iteration has made a change
+    size_t num_optimization_iters = 6;
```
can't fool clang-tidy 😀

ZolotukhinM commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -0,0 +1,208 @@
+std::unordered_map<std::string, std::shared_ptr<Graph>> operator_functions;
+
+c10::optional<size_t> normIndex(int64_t index, size_t len) {
+  if (index < 0) {
+    index = index + len;
+  }
```
Should it be a while loop? I.e. is -5 a valid index when the size is 4?

ZolotukhinM reviewed (commented):
Looks much better now, thanks!

Commit:
Update on "Initial Symbolic Shape Analysis"


I'm going to post on dev-discuss soon with a more thorough explanation of the design and advantages of this shape analysis, so I'm leaving out that for now.

There is still a ton left to do, I'm posting this initial version so we can get something on master multiple can work on. List of many remaining steps to do:

- [ ] Add symbolic shapes support
- [ ] Bind shape functions for operators in C++ 
- [ ] Make classes of operators share the same shape function (e.g. pointwise, broadcast two inputs)
- [ ] Refactor APIs 
- [ ] Only iteratively optimize shape function while a change has been made
- [ ] Expand coverage of coverage to common ops
- [ ] Add shape analysis pass on Graph that handles Ifs and Loops 
- [ ] Allow concurrent reads to the operator map
- [ ] Successive applications of same inputs to same shape function (e.g. series of pointwise ops)

For this review, I am mostly looking for comments related to the implementation of symolic_shape_analysis.cpp, with the caveats listed above. I am not really looking for comments related to api/registration/graph level analysis as those are all planned to be changed. I am fine landing this as is or waiting until necessary components of the TODOs above are finished.

Differential Revision: [D27750998](https://our.internmc.facebook.com/intern/diff/D27750998)

[ghstack-poisoned]

eellison committed:
Update on "Initial Symbolic Shape Analysis"


I'm going to post on dev-discuss soon with a more thorough explanation of the design and advantages of this shape analysis, so I'm leaving out that for now.

There is still a ton left to do, I'm posting this initial version so we can get something on master multiple can work on. List of many remaining steps to do:

- [ ] Add symbolic shapes support
- [ ] Bind shape functions for operators in C++ 
- [ ] Make classes of operators share the same shape function (e.g. pointwise, broadcast two inputs)
- [ ] Refactor APIs 
- [ ] Only iteratively optimize shape function while a change has been made
- [ ] Expand coverage of coverage to common ops
- [ ] Add shape analysis pass on Graph that handles Ifs and Loops 
- [ ] Allow concurrent reads to the operator map
- [ ] Successive applications of same inputs to same shape function (e.g. series of pointwise ops)

For this review, I am mostly looking for comments related to the implementation of symolic_shape_analysis.cpp, with the caveats listed above. I am not really looking for comments related to api/registration/graph level analysis as those are all planned to be changed. I am fine landing this as is or waiting until necessary components of the TODOs above are finished.

Differential Revision: [D27750998](https://our.internmc.facebook.com/intern/diff/D27750998)

[ghstack-poisoned]

eellison committed:
Update on "Initial Symbolic Shape Analysis"


I'm going to post on dev-discuss soon with a more thorough explanation of the design and advantages of this shape analysis, so I'm leaving out that for now.

There is still a ton left to do, I'm posting this initial version so we can get something on master multiple can work on. List of many remaining steps to do:

- [ ] Add symbolic shapes support
- [ ] Bind shape functions for operators in C++ 
- [ ] Make classes of operators share the same shape function (e.g. pointwise, broadcast two inputs)
- [ ] Refactor APIs 
- [ ] Only iteratively optimize shape function while a change has been made
- [ ] Expand coverage of coverage to common ops
- [ ] Add shape analysis pass on Graph that handles Ifs and Loops 
- [ ] Allow concurrent reads to the operator map
- [ ] Successive applications of same inputs to same shape function (e.g. series of pointwise ops)

For this review, I am mostly looking for comments related to the implementation of symolic_shape_analysis.cpp, with the caveats listed above. I am not really looking for comments related to api/registration/graph level analysis as those are all planned to be changed. I am fine landing this as is or waiting until necessary components of the TODOs above are finished.

Differential Revision: [D27750998](https://our.internmc.facebook.com/intern/diff/D27750998)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27750998).

eellison committed:
Update on "Initial Symbolic Shape Analysis"


I'm going to post on dev-discuss soon with a more thorough explanation of the design and advantages of this shape analysis, so I'm leaving out that for now.

There is still a ton left to do, I'm posting this initial version so we can get something on master multiple can work on. List of many remaining steps to do:

- [ ] Add symbolic shapes support
- [ ] Bind shape functions for operators in C++ 
- [ ] Make classes of operators share the same shape function (e.g. pointwise, broadcast two inputs)
- [ ] Refactor APIs 
- [ ] Only iteratively optimize shape function while a change has been made
- [ ] Expand coverage of coverage to common ops
- [ ] Add shape analysis pass on Graph that handles Ifs and Loops 
- [ ] Allow concurrent reads to the operator map
- [ ] Successive applications of same inputs to same shape function (e.g. series of pointwise ops)

For this review, I am mostly looking for comments related to the implementation of symolic_shape_analysis.cpp, with the caveats listed above. I am not really looking for comments related to api/registration/graph level analysis as those are all planned to be changed. I am fine landing this as is or waiting until necessary components of the TODOs above are finished.

Differential Revision: [D27750998](https://our.internmc.facebook.com/intern/diff/D27750998)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27750998).

eellison committed:
Update on "Initial Symbolic Shape Analysis"


I'm going to post on dev-discuss soon with a more thorough explanation of the design and advantages of this shape analysis, so I'm leaving out that for now.

There is still a ton left to do, I'm posting this initial version so we can get something on master multiple can work on. List of many remaining steps to do:

- [ ] Add symbolic shapes support
- [ ] Bind shape functions for operators in C++ 
- [ ] Make classes of operators share the same shape function (e.g. pointwise, broadcast two inputs)
- [ ] Refactor APIs 
- [ ] Only iteratively optimize shape function while a change has been made
- [ ] Expand coverage of coverage to common ops
- [ ] Add shape analysis pass on Graph that handles Ifs and Loops 
- [ ] Allow concurrent reads to the operator map
- [ ] Successive applications of same inputs to same shape function (e.g. series of pointwise ops)

For this review, I am mostly looking for comments related to the implementation of symolic_shape_analysis.cpp, with the caveats listed above. I am not really looking for comments related to api/registration/graph level analysis as those are all planned to be changed. I am fine landing this as is or waiting until necessary components of the TODOs above are finished.

Differential Revision: [D27750998](https://our.internmc.facebook.com/intern/diff/D27750998)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27750998).

eellison committed:
Update on "Initial Symbolic Shape Analysis"


I'm going to post on dev-discuss soon with a more thorough explanation of the design and advantages of this shape analysis, so I'm leaving out that for now.

There is still a ton left to do, I'm posting this initial version so we can get something on master multiple can work on. List of many remaining steps to do:

- [ ] Add symbolic shapes support
- [ ] Bind shape functions for operators in C++ 
- [ ] Make classes of operators share the same shape function (e.g. pointwise, broadcast two inputs)
- [ ] Refactor APIs 
- [ ] Only iteratively optimize shape function while a change has been made
- [ ] Expand coverage of coverage to common ops
- [ ] Add shape analysis pass on Graph that handles Ifs and Loops 
- [ ] Allow concurrent reads to the operator map
- [ ] Successive applications of same inputs to same shape function (e.g. series of pointwise ops)

For this review, I am mostly looking for comments related to the implementation of symolic_shape_analysis.cpp, with the caveats listed above. I am not really looking for comments related to api/registration/graph level analysis as those are all planned to be changed. I am fine landing this as is or waiting until necessary components of the TODOs above are finished.

Differential Revision: [D27750998](https://our.internmc.facebook.com/intern/diff/D27750998)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27750998).

eellison committed:
Update on "Initial Symbolic Shape Analysis"


I'm going to post on dev-discuss soon with a more thorough explanation of the design and advantages of this shape analysis, so I'm leaving out that for now.

There is still a ton left to do, I'm posting this initial version so we can get something on master multiple can work on. List of many remaining steps to do:

- [ ] Add symbolic shapes support
- [ ] Bind shape functions for operators in C++ 
- [ ] Make classes of operators share the same shape function (e.g. pointwise, broadcast two inputs)
- [ ] Refactor APIs 
- [ ] Only iteratively optimize shape function while a change has been made
- [ ] Expand coverage of coverage to common ops
- [ ] Add shape analysis pass on Graph that handles Ifs and Loops 
- [ ] Allow concurrent reads to the operator map
- [ ] Successive applications of same inputs to same shape function (e.g. series of pointwise ops)

For this review, I am mostly looking for comments related to the implementation of symolic_shape_analysis.cpp, with the caveats listed above. I am not really looking for comments related to api/registration/graph level analysis as those are all planned to be changed. I am fine landing this as is or waiting until necessary components of the TODOs above are finished.

Differential Revision: [D27750998](https://our.internmc.facebook.com/intern/diff/D27750998)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27750998).
